crates/biome_js_analyze/src/semantic_analyzers/nursery/use_sorted_classes.rs

```diff
@@ -0,0 +1,1373 @@
+    match get_attribute_name(attribute) {
+        Some(name) => attributes.contains(&name.to_string()),
+        None => false,
+    }
+}
+
+fn get_options_from_analyzer(analyzer_options: &AnalyzerOptions) -> UseSortedClassesOptions {
+    match analyzer_options
+        .configuration
+        .rules
+        .get_rule_options::<UseSortedClassesOptions>(&RuleKey::new("nursery", "useSortedClasses"))
+    {
+        Some(options) => options.clone(),
+        None => UseSortedClassesOptions::default(),
+    }
+}
+
+// class parser
+// ------------
+
+fn split_at_indexes<'a>(s: &'a str, indexes: &[usize]) -> Vec<&'a str> {
+    let mut segments = Vec::new();
+    let mut start_offset = 0;
+    let mut start = 0;
+
+    for &index in indexes {
+        if index > s.len() {
+            break; // Avoid panicking on out-of-bounds indexes
+        }
+        if index > start {
+            segments.push(&s[start + start_offset..index]);
+        }
+        start_offset = 1;
+        start = index;
+    }
+
+    if start + start_offset < s.len() {
+        segments.push(&s[start + start_offset..]);
+    }
+
+    segments
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum Quote {
+    Single,
+    Double,
+    Backtick,
+}
+
+impl Quote {
+    fn from_char(c: char) -> Option<Quote> {
+        match c {
+            '\'' => Some(Quote::Single),
+            '"' => Some(Quote::Double),
+            '`' => Some(Quote::Backtick),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum CharKind {
+    Other,
+    Quote(Quote),
+    Backslash,
+}
+
+#[derive(Debug)]
+struct UtilitySegmentData {
+    arbitrary: bool,
+    text: String,
+}
+
+#[derive(Debug)]
+struct UtilityData {
+    variants: Vec<UtilitySegmentData>,
+    utility: UtilitySegmentData,
+}
+
+fn parse_class(class_name: &str) -> UtilityData {
+    // state
+    let mut arbitrary_block_depth = 0;
+    let mut at_arbitrary_block_start = false;
+    let mut quoted_arbitrary_block_type: Option<Quote> = None;
+    let mut last_char = CharKind::Other;
+    let mut delimiter_indexes: Vec<usize> = Vec::new();
+
+    // loop
+    for (index, c) in class_name.chars().enumerate() {
+        let mut next_last_char = CharKind::Other;
+        let mut is_start_of_arbitrary_block = false;
+        match c {
+            '[' => {
+                if arbitrary_block_depth == 0 {
+                    arbitrary_block_depth = 1;
+                    at_arbitrary_block_start = true;
+                    is_start_of_arbitrary_block = true;
+                } else if let None = quoted_arbitrary_block_type {
+                    arbitrary_block_depth += 1;
+                }
+            }
+            '\'' | '"' | '`' => {
+                if at_arbitrary_block_start {
+                    let quote = Quote::from_char(c)
+                        .expect("TODO: error message (this should never happen)");
+                    quoted_arbitrary_block_type = Some(quote.clone());
```

Comment:
hmm I am worried that if I don't panic for that impossible state, errors might slip through

Reply:
We don't usually panic. If there's an error, we have three options:
- return `None` and the rule doesn't trigger
- the rule provides a different error message about incorrect classes
- a new rule that checks this invalid statw

Reply:
This is more of a "this program state is impossible" thing. I can just remove it, as I'm pretty certain it can never happen, it's just that Rust is not smart enough to realize. I usually add errors as "sanity checks" in these cases, just so that a silent error doesn't slip through.

In this case, as I'm matching those characters, `Quote::from_char` is certain to return `Some`, because it does the same check underneath.

Reply:
I removed the sanity check and the cloning.